linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala

```diff
@@ -644,18 +649,19 @@ private final class ClassDefChecker(classDef: ClassDef, reporter: ErrorReporter)
         checkTrees(elems, env)
 
       case ArrayLength(array) =>
-        if (!array.tpe.isInstanceOf[ArrayType])
-          reportError(i"Array type expected but ${array.tpe} found")
+        checkArrayReceiverType(array.tpe)
         checkTree(array, env)
 
       case ArraySelect(array, index) =>
-        if (!array.tpe.isInstanceOf[ArrayType])
-          reportError(i"Array type expected but ${array.tpe} found")
+        checkArrayReceiverType(array.tpe)
         checkTree(array, env)
         checkTree(index, env)
 
-      case _:RecordSelect | _:RecordValue =>
-        reportError("invalid tree")
+      case RecordSelect(record, _) if allowTransients =>
+        checkTree(record, env)
```

Comment:
Consider something like the following instead:
```suggestion
      case RecordSelect(record, _) =>
        if (!allowTransients)
          reportError("invalid tree")
        checkTree(record, env)
```
Then you can remove the default `case _: ... =>` at the end.

That duplicates the error message a bit, but it improves the exhaustivity check for the whole `match`, as well as keep checking subtrees even if we find a bad one.

Reply:
Good point. I ended up making a `checkAllowTransients()` which gives us the best of both worlds :)